docs/design_notes/lift.md

```diff
@@ -0,0 +1,355 @@
+
+    Now that the variables and PRNGs are lifted inside the transformation, we want to recreate the lifted scopes. Pack calls
+    `fn` with a `scope_fn` that takes the lifted variables and PRNGs and returns the reconstructed scopes with the lifted variables and rng sequences.
+
+5. *Repack stage*
+
+    After we have used the lifted scopes we have to retrieve the updated variables (PRNG sequences can simply be discarded).
+    pack passes the `repack_fn` to support this.
+    This stage is similar to stage 2 except that we only lift variables and immutable variables are ignored.
+    Immutable variables cannot be updated. Therefore, they should not be returned from the transformed function. 
+
+6. *Commit stage*
+
+    `pack` expects `fn` to return a pair where the first item will simply be returned from pack and the second item should be the repacked variables.
+    The updated variables are stored in the original/un-lifted scopes such that the mutations that happen inside the transformation survive after the transformation completes.
+
+
+### Using pack example
+
+
+A minimal example of using `pack` to transpose each matrix in a variable collection:
+
+```python
+from flax.core import lift
+from flax.core import Scope, init, apply, nn as core_nn
+
+def lift_transpose(fn, target='params', variables=True, rngs=True)
+  # by default we transpose 'params' and simply pass through all other variables.
+  def wrapper(scope_fn, repack_fn, variable_groups, rng_groups, *args):
+    # normally we would first call into a JAX transformed function here...
+    target, rest = variable_groups
+    def trans(x):
+      if x.ndim == 2:
+        return x.T
+      return x
+    target = jax.tree_map(trans, target)
+    variable_groups = (target, rest)
+    scope = scope_fn(variable_groups, rng_groups)
+    y = fn(scope, *args)
+    out_variables = repack_fn(scope)
+    return y, out_variables
+  return lift.pack(wrapper, (target, variables), (variables,), (rngs,))
+
+x = jnp.ones((3, 2))
+y, params = init(lift_id(core_nn.dense))(random.PRNGKey(0), x, 4)
+```
+
+NOTE that most users should not need to interact with `pack` directly.
+Please open a GitHub issue when you find a use case that is not supported yet by the existing lifted transformations.
+
+### Supported transformations
+
+| Jax Transform | Supported in Linen? | Comments |
+|-|-|-|
+| vmap | ✅ |  |
+| scan | ✅ | Carry variables cannot be initialized inside the scan body. |
+| remat | ✅ |  |
+| jit | ✅ | Current implementation might cause unnecessary recompilation. |
+| pmap | ❌ |  |
+| grad | ❌ |  |
+| xmap | ❌ |  |
+| custom_vjp | ❌ |  |
+| custom_jvp | ❌ |  |
+
+References:
+- [Linen transforms documentation](https://flax.readthedocs.io/en/latest/flax.linen.html#module-flax.linen.transforms).
+- [Linen transforms source code](https://github.com/google/flax/blob/master/flax/linen/transforms.py)
+- [Core lifting source code](https://github.com/google/flax/blob/master/flax/core/lift.py)
```

Comment:
Should we link back from the source to RTD?

(`lift.pack` has very terse documentation currently :-)

Reply:
Probably. I think the core docs need some more thought and love. I'll do that in a seperate PR